kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -51,13 +55,13 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
         if (msg instanceof DecodedRequestFrame) {
             DecodedRequestFrame<?> decodedFrame = (DecodedRequestFrame<?>) msg;
             // Guard against invoking the filter unexpectedly
-            if (filter.shouldDeserializeRequest(decodedFrame.apiKey(), decodedFrame.apiVersion())) {
+            if (invoker.shouldHandleRequest(decodedFrame.apiKey(), decodedFrame.apiVersion())) {
```

Comment:
Similar to your rational in https://github.com/kroxylicious/kroxylicious/pull/288/files#r1178481122 I wonder if `invoker` is an implementation detail and the business function (for want of a better term) is `dispatcher`? 

`invoker.shouldHandleRequest` reads very awkwardly to me. 

To me the natural interface of an invoker is `invoke(decodedFrame)` and we push this whole block down into the invoker itself the invoker would then ask the filters if they handle the request and invoke `onRequest` if they chose to.  It would be a similar story for `*Response`.

Reply:
yeah that'd be an improvement, calling `shouldDeserializeRequest` from in `onRequest`. Though currently we still need a `boolean shouldHandleRequest` available for making the decode decision early, the "if any Filter handles this message -> decode" logic. I'm wondering if we could, instead of deciding up front, pass around a Frame that contains the bytes and will lazily deserialize and keep a reference to the Header/ApiMessage if anything asks for them.

Reply:
Actually small snag there. `onRequest` and `onResponse`are not currently aware of the apiVersion. Maybe we should make them aware?

Reply:
This change gets tricky because the `filter-api` module (rightly) doesn't depend on the core `kroxylicious` module and thus we can't include `decodedFrame` in the invoker API. There is some debate as to which module `invoker` should exist in. 

Our current thinking is to push the branching logic down into the invoker at the expense of additional garbage creation by creating a `DefaultFilterContext` for each request.